zerver/models.py

```diff
@@ -3351,12 +3356,40 @@ def to_dict(self) -> Dict[str, Any]:
 post_delete.connect(flush_used_upload_space_cache, sender=Attachment)
 
 
+def validate_attachment_request_for_spectator_access(
+    realm: Realm, attachment: Attachment
+) -> Optional[bool]:
+    if attachment.realm != realm:
+        return False
+
+    # Update cached is_web_public property, if necessary.
+    if attachment.is_web_public is None:
+        attachment.is_web_public = Stream.objects.filter(
+            realm=realm,
+            invite_only=False,
+            is_web_public=True,
+            recipient__in=attachment.messages.values("recipient"),
+        ).exists()
+        attachment.save(update_fields=["is_web_public"])
```

Comment:
This codepath is going to be a hot path, with multiple parallel executions, when a message is moved into a stream with a large number of people.  All users will attempt to load the image, and end up in this block.

In most cases, this is merely duplicate work -- all of them will do this same query, end up at the same result, and clobber each other with the same value.

There's a potential race condition where a message gets moved private -> public -> private in a short period of time, leading to the `is_web_public` value from the middle step getting written _after_ the cache column has been cleared in the move to the private one.  The `do_update_message` code is already in a `transaction.atomic`, so doing this safely involves making sure these queries are also done in one, with a write lock when checking `is_web_public`.  Obviously, effectively serializing access to the cache is not great for performance; an optimization is probably to check the field without the lock, and SELECT FOR UPDATE it and re-check if it's still null, then do the calculation.

There's probably a deadlock potential I'm missing in here.  We can also potentially wave our hands and declare this race not significant enough to worry about right now,

Reply:
This `is_web_public` code path actually won't be a hot path for now, because we don't live-update web-public access at all, so there's no reason to expect highly synchronized access to uploaded files for web public streams (even a second of desynchronization should mean this cache isn't a race problem). But this critique applies to the `is_realm_public` path as well.

Even for `is_realm_public`, t's only going to come up when images are moved to a different stream, which is a relatively rare event, especially for happening in bulk. So I agree it's a hot path in those situations and thus probably a problem at some level (since the Tornado pushes will cause clients to do a synchronous fetch). But it may not be an actual issue in practice often, especially if our browser caching story were correct (in which case everyone who could see the images in question before the move won't do a new request for them after a move). Which I know it isn't entirely due to #16791 being unresolved.

I think this at least argues for refining the `do_update_message` logic to only set these to `null` if the two streams have different `is_web_public` / `is_realm_public` settings, since I think the vast majority of actions moving messages from one stream to another will not have that property. 

> Obviously, effectively serializing access to the cache is not great for performance; an optimization is probably to check the field without the lock, and SELECT FOR UPDATE it and re-check if it's still null, then do the calculation.

And then I think this would be a pretty reasonable optimization to attempt as well, probably starting with extracting this block to a function.

Reply:
Applied this change to avoid clearing the cache for moves between streams with the same access permissions.

```
diff --git a/zerver/lib/actions.py b/zerver/lib/actions.py
index a0cbe87d5a..a6866f2c22 100644
--- a/zerver/lib/actions.py
+++ b/zerver/lib/actions.py
@@ -7070,18 +7070,22 @@ def do_update_message(
             send_event(user_profile.realm, delete_event, delete_event_notify_user_ids)
 
             # Reset the Attachment.is_*_public caches for all messages
-            # moved to another stream.
-            #
-            # This could be optimized to only clear these caches if
-            # the new stream's permissions are different from the old.
-            Attachment.objects.filter(messages__in=changed_message_ids).update(
-                is_realm_public=None,
-                is_web_public=None,
-            )
-            ArchivedAttachment.objects.filter(messages__in=changed_message_ids).update(
-                is_realm_public=None,
-                is_web_public=None,
-            )
+            # moved to another stream with different access permissions.
+            if new_stream.invite_only != stream_being_edited.invite_only:
+                Attachment.objects.filter(messages__in=changed_message_ids).update(
+                    is_realm_public=None,
+                )
+                ArchivedAttachment.objects.filter(messages__in=changed_message_ids).update(
+                    is_realm_public=None,
+                )
+
+            if new_stream.is_web_public != stream_being_edited.is_web_public:
+                Attachment.objects.filter(messages__in=changed_message_ids).update(
+                    is_web_public=None,
+                )
+                ArchivedAttachment.objects.filter(messages__in=changed_message_ids).update(
+                    is_web_public=None,
+                )
 
     # This does message.save(update_fields=[...])
     save_message_for_edit_use_case(message=target_message)
diff --git a/zerver/tests/test_subs.py b/zerver/tests/test_subs.py
index fb922d8a91..07f7ce695c 100644
--- a/zerver/tests/test_subs.py
+++ b/zerver/tests/test_subs.py
@@ -1044,8 +1044,9 @@ class StreamAdminTest(ZulipTestCase):
         stream.refresh_from_db()
         self.assertFalse(attachment.is_web_public)
 
-        # Verify that editing a message does the right thing as well.
+        # Verify moving a message to another public stream doesn't reset cache.
         new_stream = self.make_stream("new_stream", realm=realm)
+        self.subscribe(owner, "new_stream")
         result = self.client_patch(
             "/json/messages/" + str(msg_id),
             {
@@ -1053,10 +1054,49 @@ class StreamAdminTest(ZulipTestCase):
                 "propagate_mode": "change_all",
             },
         )
+        self.assert_json_success(result)
+        attachment.refresh_from_db()
+        self.assertFalse(attachment.is_web_public)
+        self.assertTrue(attachment.is_realm_public)
+
+        # Verify moving a message to a private stream
+        private_stream = self.make_stream("private_stream", realm=realm, invite_only=True)
+        self.subscribe(owner, "private_stream")
+        result = self.client_patch(
+            "/json/messages/" + str(msg_id),
+            {
+                "stream_id": private_stream.id,
+                "propagate_mode": "change_all",
+            },
+        )
+        self.assert_json_success(result)
+        attachment.refresh_from_db()
+        self.assertFalse(attachment.is_web_public)
+        self.assertIsNone(attachment.is_realm_public)
+
+        self.assertFalse(validate_attachment_request(cordelia, attachment.path_id))
+        self.assertTrue(validate_attachment_request(owner, attachment.path_id))
+        attachment.refresh_from_db()
+        self.assertFalse(attachment.is_realm_public)
+
+        # Verify moving a message to a web-public stream
+        web_public_stream = self.make_stream("web_public_stream", realm=realm, is_web_public=True)
+        result = self.client_patch(
+            "/json/messages/" + str(msg_id),
+            {
+                "stream_id": web_public_stream.id,
+                "propagate_mode": "change_all",
+            },
+        )
+        self.assert_json_success(result)
         attachment.refresh_from_db()
         self.assertIsNone(attachment.is_web_public)
         self.assertIsNone(attachment.is_realm_public)
 
+        self.assertTrue(validate_attachment_request_for_spectator_access(realm, attachment))
+        attachment.refresh_from_db()
+        self.assertTrue(attachment.is_web_public)
+
     def test_try_make_stream_public_with_private_history(self) -> None:
         user_profile = self.example_user("hamlet")
         self.login_user(user_profile)

```

Reply:
Totally untested, but something like:
```suggestion
    if attachment.is_web_public is None:
        Attachment.objects.filter(
            id=atachment.id, is_web_public__isnull=True
        ).update(
            is_web_public=Exists(
                message_model.objects.filter(
                    attachment=OuterRef("id"),
                    recipient__stream__invite_only=False,
                    recipient__stream__is_web_public=True,
                ),
            ),
        )
        attachment. refresh_from_db()
```

Reply:
Applied this edit both here and for is_realm_public, after fixing the trivial typos and variable name issues. Tests pass after the change, and I like the test suite for this block of, so I think it's good.